### PR TITLE
feat: tag notes index

### PR DESCRIPTION
**Pull request by moughxyz**

### CODE REVIEW

File: packages/snjs/lib/protocol/collection/tag_notes_index.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+  private recieveTagChanges(tags: SNTag[]): void {
+    for (const tag of tags) {
+      const uuids = tag.noteReferences.map((ref) => ref.uuid);
+      const countableUuids = uuids.filter((uuid) =>
+        this.allCountableNotes.has(uuid)
+      );
+      this.tagToNotesMap[tag.uuid] = new Set(countableUuids);
+    }
+  }
```
Comment (laurentsenta) [inline packages/snjs/lib/protocol/collection/tag_notes_index.ts:45]: a few typos, you could search and replace recieve -> receive

File: packages/snjs/lib/services/item_manager.ts [hunk 2]
```diff
@@ -313,6 +324,13 @@ export class ItemManager extends PureService {
       this.collection.discard(item);
     }
     this.notesView.setNeedsRebuilding();
+    this.tagNotesIndex.receiveTagAndNoteChanges(
+      changedOrInserted.filter(
+        (item) =>
+          item.content_type === ContentType.Tag ||
+          item.content_type === ContentType.Note
+      ) as SNTag[]
+    );
```
Comment (laurentsenta) [inline packages/snjs/lib/services/item_manager.ts:333]: It's probably `as SNTag | SNNote` here.

File: packages/snjs/lib/protocol/collection/tag_notes_index.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+  public countableNotesForTag(tag: SNTag): number {
+    return this.tagToNotesMap[tag.uuid]?.size;
+  }
+
+  public receiveTagAndNoteChanges(items: (SNTag | SNNote)[]): void {
+    const notes = items.filter(
+      (item) => item.content_type === ContentType.Note
+    ) as SNNote[];
```
Comment (laurentsenta) [inline packages/snjs/lib/protocol/collection/tag_notes_index.ts:28]: With a typeguard you should get the correct type without static casts,

`const isNote = (x: SNItem): x is SNNote => x.content_type === ContentType.Note`

Same for tags, and the call to receiveTagAndNoteChanges in item manager:

```
    // should work and be typesafe.
    const notes: SNote[] = items.filter(isNote );
```

Comment (moughxyz) [reply packages/snjs/lib/protocol/collection/tag_notes_index.ts:28]: Good call 👍

**Review by laurentsenta — APPROVED** [review]

**Pull request merged**